test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java

```diff
@@ -88,10 +96,120 @@ public void testEncoding() throws Exception {
+                    -66, -126, -27, -128, -92, -17, -66, -126, -27, -90, -91, -17, -67, -91, -17, -66, -126, -27, -126, -84,
+                    -17, -67, -93, -17, -66, -126, -27, -72, -126 };
+    
+            String value = new String(data, StandardCharsets.UTF_8);
+    
+            HTTPArgument arg;
+            // Passing real encoded value of Japanese character as
+            // %8F%89%8A%FA%92l%91%E5%8D%E3%8Es and try to decode with Shift_JIS encoding
+            arg = new HTTPArgument("name.?", encodedValue, true, "Shift_JIS");
+            assertEquals("arg.getValue() should be equal to " + value, value, arg.getValue());
+        }
+    
+        @Test
+        public void testWithoutFix() throws Exception {
+    
+            String encodedValue = "%8F%89%8A%FA%92l%91%E5%8D%E3%8Es";
+    
+            byte[] data = new byte[] { -17, -66, -126, -26, -72, -119, -27, -113, -81, -25, -108, -104, -17, -67, -70, -17,
+                    -66, -126, -27, -128, -92, -17, -66, -126, -27, -90, -91, -17, -67, -91, -17, -66, -126, -27, -126, -84,
+                    -17, -67, -93, -17, -66, -126, -27, -72, -126 };
+    
+            String value = new String(data, StandardCharsets.UTF_8);
+    
+            HTTPArgument arg;
+            // Passing real encoded value of Japanese character as
+            // %8F%89%8A%FA%92l%91%E5%8D%E3%8Es and try to decode with UTF-8 encoding
+            arg = new HTTPArgument("name.?", encodedValue, true, "UTF-8");
+            assertNotEquals("arg.getValue() should not be equal to " + value, value, arg.getValue());
+        }
+    
+        private void testEncodings(String encoding) {
+            HTTPArgument arg;
+            arg = new HTTPArgument("name.?", "value_ here", false, encoding);
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getValue());
+            assertEquals("arg.getEncodedName() should be equal to name.%3F", "name.%3F", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_+here", "value_+here", arg.getEncodedValue());
+            // Show that can bypass encoding:
+            arg.setAlwaysEncoded(false);
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getEncodedValue());
+    
+            // When the name and value parameter does not have valid encoded value (Encoded
+            // and Decoded values are same)
+            // In real encoded value for "name.?" is "name.%3F" and the encoded value for
+            // "value_here" is "value_+here"
+            arg = new HTTPArgument("name.?", "value_ here", true, encoding);
+    
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
```

Comment:
It is great you add assert messages, however it makes little sense to duplicate "expected" value in the message.

In case of failure, the message would include "expected" value anyway, so phrases like `should be equal to name.?` add nothing.

You should probably use something like `assertEquals("HTTPArgument(\"name.?\", \"value_ here\",...).getName()", "name.?", arg.getName());`

Then it would produce clear message and tell what was the input, the action, the expected output and the actual output.

Reply:
Assert messages have been updated as per your suggestion.